osresearch/heads
# libgfxinit/nativegfx init: efifb enforced fb (+coreboot ramstage enabled bootsplash)

EDIT: simplefb support (previous effort) was replaced by efifb for simplicity and possible generalization of usage across FSP GOP/native gfx init/libgfxinit under coreboot/heads usage as a payload under linuxboot (without added compexity of hacking kexec, have coreboot taint kernel on linux command line options ot Heads payload).

- This frees 500kb of firmware space, as detailed in following comments, removing DRM+FB drivers of the kernel.
- It also provides a common base to support GOP enabled graphcis boards, while still untested on my side.
- It also permits to have splashscreen that are smaller then board defined HEIGHT/WIDTH under coreboot config, centering that bootsplash correctly.
- **dGPU fb is now linear, and enforces bootsplash from optionrom fb setuped for 1280x1024** which needs testing

Bonus it works also on qemu target boards (bochs native graphic init + libgfxinit).
Leaving traces below for history.Changes qemu call from -vga virtio to -vga std (to use bochs coreboot native graphic init + fb)



efifb discussions and changes from https://github.com/osresearch/heads/pull/1403#issuecomment-1636219081 and on


Testing needed and **why**:
- [x] x230-legacy/x230-legacy-flash: @tlaurion :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** fixed from https://github.com/osresearch/heads/pull/1403#issuecomment-1677476841
- [ ] t430-legacy/t430-legacy-flash: @daringer : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** fixed from https://github.com/osresearch/heads/pull/1403#issuecomment-1677476841
- [x] t430-maximized : @daringer : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** fixed from https://github.com/osresearch/heads/pull/1403#issuecomment-1677476841 Thanks to @srgrint for last minute testing which saved t430-maximized and t430-hotp-maximized from going untested prior of merging this PR.

----

- [x] x220 (xx20): @Thrilleratplay @BlackMaria @srgrint : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** Thanks @srgrint for testing!
- [x] [t420](https://doc.coreboot.org/mainboard/lenovo/t420.html) (xx20): @alexmaloteaux @natterangell (iGPU) @akfhasodh @doob85 : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** Thanks to [natterangell](https://github.com/natterangell) for testing!
- [ ] [t430](https://doc.coreboot.org/mainboard/lenovo/t430.html) (xx30): @nestire(t430-legacy, t430-maximized) @Thrilleratplay @alexmaloteaux @lsafd @bwachter(iGPU maximized) @shamen123 @eganonoa(iGPU) @nitrosimon @jans23 @icequbes1 (iGPU) @weyounsix (t430-dgpu) @SvenSemmler (t430-hotp-maximized) : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** Thanks to @srgrint for last minute testing which saved t430-maximized and t430-hotp-maximized from going untested prior of merging this PR.
- [ ] [t430s](https://www.coreboot.org/Board:lenovo/t430s) and https://github.com/osresearch/heads/pull/1019 (xx30): @Siproqu : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [ ] [t520](https://doc.coreboot.org/mainboard/lenovo/t520.html) (xx30): NOBODY : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [ ] [t530](https://doc.coreboot.org/mainboard/lenovo/t530.html) (xx30): @3hhh : **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [ ] [t530-**dgpu**](https://doc.coreboot.org/mainboard/lenovo/t530.html) (xx30): @3hhh : **optrom efifb prepared fb 1280x1024 + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [ ] [w530](https://doc.coreboot.org/mainboard/lenovo/w530.html) (xx30): @eganonoa @zifxify @weyounsix (dGPU: w530-k2000m) @jnscmns (dGPU K1000M) @computer-user123 (w530 / & w530 k2000 : prefers iGPU) :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** 
- [ ] [w530-**dgpu**](https://doc.coreboot.org/mainboard/lenovo/w530.html) (xx30): @eganonoa @zifxify @weyounsix (dGPU: w530-k2000m) @jnscmns (dGPU K1000M) @computer-user123 (w530 / & w530 k2000 : prefers iGPU) : **optrom efifb prepared fb 1280x1024 + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [x] X230i (x230): @natterangell :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** considered equivalent of x230. Not tested but should not be considered independent board.
- [x] x230-fhd/edp variant: @n4ru @computer-user123 (nitro caster board) @Tonux599 @househead @pcm720 (eDP 4.0 board and 1440p display) :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** TESTED by [pcm720](https://github.com/pcm720) thanks!
- [x] [t440p](https://doc.coreboot.org/mainboard/lenovo/t440p.html): @ThePlexus @srgrint @akunterkontrolle @rbreslow :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)**
- [x] [ASUS P8Z77 M PRO (Ivy bridge)](https://doc.coreboot.org/mainboard/asus/p8z77-m_pro.html): @ThePlexus :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** : thanks @srgrint for testing. Thanks @ThePlexus for testing.
- [x] https://github.com/osresearch/heads/pull/1399: @d-wid :  **libgfxinit efifb prepared fb + linux efifb + centered 1024x768 bootsplash (3 second boottime reduction)** thanks @d-wid for testing!
------
OLD (simplefb related)

https://github.com/osresearch/heads/commit/537cc831773509dde3ac4add060bb1636bd8ef59 provides the changes needed to test this PR against qemu-coreboot-whiptail-tpm1 and qemu-coreboot-fbwhiptail-tpm1 on fb(no drm/dri)

Note: doesn't work properly on qemu of debian-11. Will edit once built on top of debian-12, which latest commits should permit to build without problem and then be testable on debian-12 qemu. Will update.

This gives the same current result as on x230 when enabling simplefb alone, without all drm+gpu drivers in kernel.

---

Excerpt of above commit log:
- Changes qemu call from -vga virtio to -vga std (to use bochs coreboot native graphic init + fb)
- coreboot configs:
  - add CONFIG_FW_CONFIG=y to expose coreboot tables so fbsimple can fire up the console
  - add CONFIG_BOOTSPLASH_FILE="@BLOB_DIR@/bootsplash-1024x768.jpg" (hard to generate jpg, another story)
  - fixate CONFIG_DRIVERS_EMULATION_QEMU_BOCHS_XRES=1024 and CONFIG_DRIVERS_EMULATION_QEMU_BOCHS_YRES=768
- linux configs:
  - CONFIG_X86_SYSFB=y to have linux expect FB inititalized by firmware (EFI/BIOS: all the same requirement)
  - Removed DRM+GPU specific options # CONFIG_DRM is not set
  - add CONFIG_FB_SIMPLE=y so that linux uses coreboot's exposed FB
  - add CONFIG_FIRMWARE_EDID=y so that firmware is probed for config
  - add CONFIG_FB_MODE_HELPERS=y CONFIG_FB_TILEBLITTING=y so that fbwhiptail can deal with the framebuffer
  - add CONFIG_FRAMEBUFFER_CONSOLE_ROTATION=y CONFIG_FRAMEBUFFER_CONSOLE_DEFERRED_TAKEOVER=y to ease usage if whatever
- kexec patch : add "simple" fb support for VLFB, so that address of Heads fb is passed on kexec call

Limitations:
- BUGS:
 - qemu on debian-11 (not KVM) is not refreshing screen unless zooming in/out or chaning input (bug in qemu?)
 - FBWHIPTAIL requires VSYNC ioctl. Should not fail if not present to stay under fbwhiptail, not corrupting output.

---
Needs cleaning. Dumping here so someone reproduces my results, mostly the limitations observed here and for a patch to be made on top of fbwhiptail so that it doesn fail on VSYNC ioctl call.
